adapters/repos/db/queue/scheduler.go

```diff
@@ -0,0 +1,490 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package queue
+
+import (
+	"context"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/adapters/repos/db/vector/common"
+	enterrors "github.com/weaviate/weaviate/entities/errors"
+)
+
+type Scheduler struct {
+	SchedulerOptions
+
+	queues struct {
+		sync.Mutex
+
+		m map[string]*queueState
+	}
+
+	// context used to close pending tasks
+	ctx      context.Context
+	cancelFn context.CancelFunc
+
+	activeTasks *common.SharedGauge
+
+	wg        sync.WaitGroup
+	chans     []chan Batch
+	triggerCh chan chan struct{}
+}
+
+type SchedulerOptions struct {
+	Logger logrus.FieldLogger
+	// Number of workers to process tasks. Defaults to the number of CPUs - 1.
+	Workers int
+	// The interval at which the scheduler checks the queues for tasks. Defaults to 1 second.
+	ScheduleInterval time.Duration
+	// Function to be called when the scheduler is closed
+	OnClose func()
+}
+
+func NewScheduler(opts SchedulerOptions) *Scheduler {
+	if opts.Logger == nil {
+		opts.Logger = logrus.New()
+	}
+	opts.Logger = opts.Logger.WithField("component", "queue-scheduler")
+
+	if opts.Workers <= 0 {
+		opts.Workers = max(1, runtime.GOMAXPROCS(0)-1)
+	}
+
+	if opts.ScheduleInterval == 0 {
+		it, err := time.ParseDuration(os.Getenv("QUEUE_SCHEDULER_INTERVAL"))
+		if err != nil {
+			it = 1 * time.Second
+		}
+		opts.ScheduleInterval = it
+	}
+
+	s := Scheduler{
+		SchedulerOptions: opts,
+		activeTasks:      common.NewSharedGauge(),
+	}
+	s.queues.m = make(map[string]*queueState)
+	s.triggerCh = make(chan chan struct{})
+
+	return &s
+}
+
+func (s *Scheduler) RegisterQueue(q Queue) {
+	s.queues.Lock()
+	defer s.queues.Unlock()
+
+	s.queues.m[q.ID()] = newQueueState(q)
+
+	q.Metrics().Registered(q.ID())
+}
+
+func (s *Scheduler) UnregisterQueue(id string) {
+	s.PauseQueue(id)
+
+	q := s.getQueue(id)
+	if q == nil {
+		return
+	}
+
+	// wait for the workers to finish processing the queue's tasks
+	q.activeTasks.Wait()
```

Comment:
still don't know how `activeTasks` are increased but I'm wondering if this steps will be completed if there is a continuous enqueuing

Reply:
Pausing the queue will prevent the scheduler from dequeuing it, not enqueuing. In other terms, it's possible to enqueue while the queue is paused.
`q.activeTasks` only keeps track of the tasks that have already been dispatched to the workers, so calling `Wait` eventually returns event with continuous enqueuing